lsstp/lsstp.8
.\" Copyright 2020 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH LSSTP 8 "Jul 2020" "s390-tools" "Linux Administrator's Manual"
.SH NAME
.B "lsstp "
\- Show STP configuration information
.SH SYNOPSIS
.BI "lsstp "

.SH DESCRIPTION
.B lsstp
displays information about the current Server Time Protocol (STP) configuration
like coordinated time network (CTN) ID, timing state and leap seconds.
.SH OUTPUT
.TP
.B STP online
Indication of the online state
.TP
.B CTN ID
The ID of the coordinated time network. If it can be decoded as EBCDIC it is shown as an EBCDIC string, otherwise a hexadecimal representation is shown.
.TP
.B CTN Type
The type of timing network.
.IP
.B No CTN
STP is not configured for attachment to a CTN.
.IP
.B STP-only
STP is configured and attached to a CTN with only STP nodes.
.IP
.B Mixed
STP is configured and attached to a CTN with both STP and external time reference (ETR) nodes.
.TP
.B Stratum
The number of servers in the timing path between the local STP clock and the selected primary time server.
.TP
.B Timing mode
.IP
.B Local
The Time-of-day (TOD) clock is stepped by the local hardware oscillator and is not steered by the STP facility.
.IP
.B ETR
The TOD clock is synchronized with an attached 9037 Sysplex Timer.
.IP
.B STP
The TOD clock is steered by the STP facility to maintain synchronization with a Coordinated Server Time (CST).
.IP
.B Uninitialized
The TOD clock is not initialized. The STP facility is allowed to perform a step adjustment to the TOD clock for synchronization.
.TP
.B Timing state
The synchronization state of the STP facility. Can be unsynchronized, synchronized or stopped.
.TP
.B DST offset
The daylight savings time offset relative to UTC in minutes.
.TP
.B Timezone offset
The offset of the local time relative to UTC in minutes.
.TP
.B Time offset
The total time offset at the server. This field is only valid in mixed CTN configurations.
.TP
.B Active leap seconds
The number of leap seconds that are currently in effect at the STP facility.
.TP
.B Leap second <insertion|deletion> at
If a leap second insertion or deletion is scheduled in the STP facility, this field shows the day and time of the scheduled change.
.SH OPTIONS
.TP
.BI "-v|--version"
Print version number.
.TP
.BI "-h|--help"
Print usage text.

.SH AUTHORS
Sven Schnelle <svens@linux.ibm.com>


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------
* __v2.15.x (2020-xx-xx)__

  For Linux kernel version: 5.x

  Changes of existing tools:

  Bug Fixes:


* __v2.15.0 (2020-10-15)__

  For Linux kernel version: 5.9

  Add new tool:
  - lsstp: A small utility to display the Server Time Protocol (STP) information present in sysfs

  Changes of existing tools:
  - dumpconf: support NVMe dump/reipl device
  - ipl_tools: support clear attribute for nvme re-IPL
  - zcrypt: Support new config state with lszcrypt and chzcrypt
  - zkey: Add support for key management system plugins
        including the KMS commands:
            bind, unbind, info, configure, rencipher, list, import, refresh
  - zkey: Add EKMFWeb support to remotely generate secure keys
  - libekmfweb: Add new EKMFWeb client library
  - libutil: Add util_file_read_va()
  - libutil: Add util_file_read_i()/util_file_read_ui()

  Bug Fixes:
  - cpumf: Fix version and help printout when CPUMF is not installed
  - ziomon/ziorep_printers: fix virtual adapter CSV output
  - zipl: Fix Error when title is not the first field in BLS file


* __v2.14.0 (2020-08-21)__

  For Linux kernel version: 5.7 / 5.8

  Changes of existing tools:
  - cpacfstats: Add ECC counters
  - dbginfo: Added collection of /proc/softirqs
  - ipl-tools: Add nvme device support to lsreipl/chreipl
  - zdsfs: Add coordinated read access
  - libzds: Add curl interface to access zosmf rest api
  - util_opt: Change util_opt_init() to honor current command, if set
  Bug Fixes:
  - lsluns: Try harder to find udevadm
  - mon_tools: Update udevadm location
  - zipl: Fix NVMe partition and base device detection
  - zipl/stage3: Correctly handle diag308 response code
  - znetconf: Introduce better ways to locate udevadm

* __v2.13.0 (2020-05-06)__

  For Linux kernel version: 5.5 / 5.6

  Add new tool:
  - genprotimg: Add genprotimg to create protected virtualization images
  - genprotimg: Add sample script to verify host keys

  Changes of existing tools:
  - dbginfo: Gather bridge related data (using 'bridge')
  - dbginfo: Removed collection of /var/log/opencryptoki/
  - dbginfo: collect softnet_stat
  - dbginfo: gather ethtool output for per-queue coalescing
  - ipl_tools: Support clear attribute for FCP and CCW re-IPL
  - zdev: Report FC Endpoint Security of zfcp devices
  - zdev/dracut/95zdev/module-setup.sh: Add ctcm kernel module
  - cpumf/data: Add new deflate counters for IBM z15
  - zkey: Add support for EP11 secure keys
  - zpcictl: Initiate recover after reset
  - zipl: Add support for NVMe devices
  - zipl: A multitude of code and stability improvements

  Bug Fixes:
  - zipl: Prevent endless loop during IPL
  - zipl/libc: Fix potential buffer overflow in printf
  - zkey: Fix listing of keys on file systems reporting DT_UNKNOWN
  - zkey: Fix display of clear key size for XTS, CCA-AESCIPHER, and EP11-AES XTS keys


* __v2.12.0 (2019-12-17)__

  For Linux kernel version: 5.4

  Changes of existing tools:
  - dbginfo: Gather qdisc related data (using 'tc')
  - dbginfo: Gather extended network statistics (using 'ip link')
  - dbginfo: Collect all files under /usr/lib/systemd/system/
  - cpumf/cpumf_helper: Add IBM z15 machine name
  - zkey: Display MKVP when validating a secure key
  - zkey: Cross check APQNs when generating, validating, or importing secure keys,
      and when changing APQN associations
  - zkey: Check crypto card level during APQN cross checking
  - zkey: Add support for generating, validating, and re-enciphering AES CIPHER keys
  - zkey-cryptsetup: Add --to-new and --from-old options
  - zkey-cryptsetup: Allow setkey to set different key types
  - lszcrypt/chzcrypt: CEX7S exploitation support
  - zcryptstats: Add support for CEX7 crypto card
  - zipl: Ship a minimal zipl.conf
  - zipl: Add value of target= as search path for BLS case

  Bug Fixes:
  - dasdview: Fix exit status in error cases
  - zipl: Fix various compile warnings
  - zipl: Fix dependency generation in zipl/boot
  - zipl: Fix entry point for stand-alone kdump
  - zipl: Add missing options to help output

* __v2.11.0 (2019-09-06)__

  For Linux kernel version: 5.3

  Changes of existing tools:
  - dasdfmt: Add support for thin-provisioned volumes
  - lsdasd: Add support for thin-provisioned volumes
  - libdasd: Provide function to utilise release space ioctl
  - libdasd: Provide function to read ese sysfs attribute
  - dbginfo: Add lspci (PCI devices) and smc_dbg (SMC sockets)
  - dbginfo: Gather ethtool related data

  Bug Fixes:
  - zipl: Fix freeing of uninitialized pointer
  - zipl: Set correct secure IPL default value

* __v2.10.0 (2019-07-31)__

  For Linux kernel version: 5.2

  Changes of existing tools:
  - zdev: Add zfcp dix parameter handling
  - cpumf: Add support for CPU-Measurement Facility counters SVN 6

  Bug Fixes:
  - libutil: Add functions to test path is read/write-only
  - zdev: Fix reporting of read-only sysfs attributes
  - zdev: Improve handling of invalid udev rules
  - zipl: Fix stfle zero padding
  - zipl: Fix build issues
  - zipl: Remove trailing spaces from the fields defined in BLS files
  - zipl: Do not overwrite BOOT_IMAGE entry
  - zkey: Fix auto-detection of clear key bitsize for XTS keys

* __v2.9.0 (2019-05-21)__

  For Linux kernel version: 5.0 / 5.1

  Add new tool:
  - zcryptstats: Add zcryptstats to display usage statistics of
      IBM Crypto Express adapters

  Changes of existing tools:
  - lszfcp: New command line option to show module parameters
  - lszfcp: Sdev attributes for scsi_disk, block, integrity, queue, iosched
  - lszfcp: Add new output marker for non-good SCSI devices (luns)
  - lszfcp: Add new output marker for non-good fc_rports
  - lszfcp: Clean up whitespace (mixed indentation, trailing)
  - lschp: Add support for specifying a CHPID
  - zipl: Add secure boot capabilities
  - zkey: Add common passphrase options for cryptsetup and crypttab

lsstp/lsstp.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <iconv.h>
#include <inttypes.h>

#include "lib/util_opt.h"
#include "lib/util_file.h"
#include "lib/util_prg.h"
#include "lib/util_path.h"

static const struct util_prg prg = {
	.desc = "Display STP system information",
	.args = "",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 2020,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

static struct util_opt opt_vec[] = {
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	UTIL_OPT_END
};

struct stp_parms {
	uint64_t ctn_id;
	unsigned int online;
	unsigned int leap_seconds;
	int leap_seconds_diff;
	unsigned int leap_seconds_utc;
	unsigned int stratum;
	unsigned int ctn_type;
	unsigned int timing_mode;
	unsigned int timing_state;
	int dst_offset;
	int time_offset;
	int time_zone_offset;
};

static int convert_ctn_id(char *in, char *out)
{
	iconv_t ic;
	size_t inlen = sizeof(unsigned long long);
	size_t outlen = sizeof(unsigned long long);

	ic = iconv_open("ISO-8859-1", "EBCDIC-US");
	if (ic == (iconv_t)-1) {
		warn("Could not initialize EBCDIC to ISO-8859-1 conversion table");
		return -1;
	}

	if (iconv(ic, &in, &inlen, (char **)&out, &outlen) == (size_t)-1) {
		warn("Code page translation EBCDIC to ISO-8859-1 failed");
		iconv_close(ic);
		return -1;
	}
	iconv_close(ic);
	return 0;
}

static const char *ctn_type_str(int type)
{
	switch (type) {
	case 0:
		return "No CTN defined";
	case 1:
		return "STP-only";
	case 2:
		return "mixed";
	default:
		return "unknown";
	}
}

static const char *tmd_to_str(int mode)
{
	switch (mode) {
	case 0:
		return "Local";
	case 1:
		return "ETR";
	case 2:
		return "STP";
	case 15:
		return "Uninitialized";
	default:
		return "unknown";
	}
}

static const char *tst_to_str(int mode)
{
	switch (mode) {
	case 0:
		return "Unsynchronized";
	case 1:
		return "Synchronized";
	case 2:
		return "Physical clock stopped";
	default:
		return "unknown";
	}
}

static const char *yesno_str(int val)
{
	return val ? "yes" : "no";
}

#define read_sysfs_attr(attr, parm, func, base)							\
	do {											\
		path = util_path_sysfs("devices/system/stp/%s", attr);				\
		ret = func(parm, base, path);							\
		if (ret) {									\
			fprintf(stderr, "failed to open %s: %s\n", path, strerror(errno));	\
			free(path);								\
			exit(EXIT_FAILURE);							\
		}										\
		free(path);									\
	} while (0)

int main(int argc, char **argv)
{
	struct stp_parms parm = { 0 };
	char ctn_id[32] = { 0 };
	char *path;
	int ret, c;

	util_prg_init(&prg);
	util_opt_init(opt_vec, NULL);

	for (;;) {
		c = util_opt_getopt_long(argc, argv);
		if (c == -1)
			break;

		switch (c) {
		case 'v':
			util_prg_print_version();
			exit(EXIT_SUCCESS);
		case 'h':
			util_prg_print_help();
			util_opt_print_help();
			exit(EXIT_SUCCESS);
		default:
			fprintf(stderr, "Try 'lsstp --help' for more information.\n");
			exit(EXIT_FAILURE);
		}
	}

	read_sysfs_attr("online", &parm.online, util_file_read_ui, 10);
	if (!parm.online) {
		printf("STP disabled\n");
		goto out;
	}

	read_sysfs_attr("ctn_id", &parm.ctn_id, util_file_read_ul, 16);
	read_sysfs_attr("ctn_type", &parm.ctn_type, util_file_read_ui, 10);
	read_sysfs_attr("stratum", &parm.stratum, util_file_read_ui, 10);
	read_sysfs_attr("leap_seconds", &parm.leap_seconds, util_file_read_ui, 10);
	read_sysfs_attr("timing_mode", &parm.timing_mode, util_file_read_ui, 10);
	read_sysfs_attr("timing_state", &parm.timing_state, util_file_read_ui, 10);
	read_sysfs_attr("dst_offset", &parm.dst_offset, util_file_read_i, 10);
	read_sysfs_attr("time_offset", &parm.time_offset, util_file_read_i, 10);
	read_sysfs_attr("time_zone_offset", &parm.time_zone_offset, util_file_read_i, 10);

	if (convert_ctn_id((char *)&parm.ctn_id, ctn_id))
		snprintf(ctn_id, sizeof(ctn_id)-1, "%016" PRIx64, parm.ctn_id);

	printf("STP online:            %s\n"
	       "CTN ID:                %s\n"
	       "CTN type:              %s\n"
	       "Stratum:               %d\n"
	       "Timing mode:           %s\n"
	       "Timing state:          %s\n"
	       "DST offset:            %d\n"
	       "Timezone offset:       %d\n"
	       "Time offset:           %d\n"
	       "Active leap seconds:   %d\n",
	       yesno_str(parm.online),
	       ctn_id,
	       ctn_type_str(parm.ctn_type),
	       parm.stratum,
	       tmd_to_str(parm.timing_mode),
	       tst_to_str(parm.timing_state),
	       parm.dst_offset,
	       parm.time_zone_offset,
	       parm.time_offset,
	       parm.leap_seconds);

	printf("Scheduled leap second: ");

	path = util_path_sysfs("devices/system/stp/leap_seconds_scheduled");
	if (util_file_read_va(path, "%d,%d", &parm.leap_seconds_utc,
				&parm.leap_seconds_diff) == 2 &&
			parm.leap_seconds_diff && parm.leap_seconds_utc) {
		time_t lsoup = parm.leap_seconds_utc;

		printf("%s at: %s UTC",
			parm.leap_seconds_diff > 0 ? "insertion" : "deletion",
			ctime(&lsoup));
	} else {
		printf("-\n");
	}
	free(path);
	return 0;
out:
	return 1;
}
